src/libs/DistanceRequestUtils.ts

```diff
@@ -301,6 +314,52 @@ function getTaxableAmount(policy: OnyxEntry<Policy>, customUnitRateID: string, d
     return amount * taxClaimablePercentage;
 }
 
+function getDistanceUnit(transaction: OnyxEntry<Transaction>, mileageRate: OnyxEntry<MileageRate>) {
+    return transaction?.comment?.customUnit?.distanceUnit ?? mileageRate?.unit ?? CONST.CUSTOM_UNITS.DISTANCE_UNIT_MILES;
+}
+
+/**
+ * Get the selected rate for a transaction, from the policy or P2P default rate.
+ * Use the distanceUnit stored on the transaction by default to prevent policy changes modifying existing transactions. Otherwise, get the unit from the rate.
+ */
+function getRate({
+    transaction,
+    policy,
+    policyDraft,
+    useTransactionDistanceUnit = true,
+}: {
+    transaction: OnyxEntry<Transaction>;
+    policy: OnyxEntry<Policy>;
+    policyDraft?: OnyxEntry<Policy>;
+    useTransactionDistanceUnit?: boolean;
+}): MileageRate {
+    let mileageRates = getMileageRates(policy, true, transaction?.comment?.customUnit?.customUnitRateID);
+    if (isEmptyObject(mileageRates) && policyDraft) {
+        mileageRates = getMileageRates(policyDraft, true, transaction?.comment?.customUnit?.customUnitRateID);
+    }
+    const policyCurrency = policy?.outputCurrency ?? PolicyUtils.getPersonalPolicy()?.outputCurrency ?? CONST.CURRENCY.USD;
+    const defaultMileageRate = getDefaultMileageRate(policy);
+    const customUnitRateID = TransactionUtils.getRateID(transaction) ?? '';
+    const mileageRate = TransactionUtils.isCustomUnitRateIDForP2P(transaction) ? getRateForP2P(policyCurrency, transaction) : mileageRates?.[customUnitRateID] ?? defaultMileageRate;
+    const unit = getDistanceUnit(useTransactionDistanceUnit ? transaction : undefined, mileageRate);
+    return {
+        ...mileageRate,
+        unit,
+        currency: mileageRate.currency ?? policyCurrency,
+    };
+}
+
+/**
+ * Get the updated distance unit from the selected rate instead of the distanceUnit stored on the transaction.
+ * Useful for updating the transaction distance unit when the distance or rate changes.
+ *
+ * For example, if an expense is '10 mi @ $1.00 / mi' and the rate is updated to '$1.00 / km',
+ * then the updated distance unit should be 'km' from the updated rate, not 'mi' from the currently stored transaction distance unit.
+ */
+function getUpdatedDistanceUnit({transaction, policy, policyDraft}: {transaction: OnyxEntry<Transaction>; policy: OnyxEntry<Policy>; policyDraft?: OnyxEntry<Policy>}) {
+    return getRate({transaction, policy, policyDraft, useTransactionDistanceUnit: false}).unit;
+}
```

Comment:
When we change the rate, we'd be always using the rate from the policy right? In that case can we just not pass the transaction?

In P2P distance requests I don't see the option to change the rate

Reply:
Correct, the rate always comes from the policy in this function. The transaction is used in multiple places within getRate such as to find the selected customUnitRateID, so it would be challenging without passing it. Like you said, I pass the transaction as undefined within getRate here if the distance unit shouldn't come from the transaction.

https://github.com/Expensify/App/blob/00eeb4a8c00a29e39d398d438ed8e16f2de08908/src/libs/DistanceRequestUtils.ts#L344